lib/ansible/vars/manager.py

```diff
@@ -661,3 +674,20 @@ def set_host_variable(self, host, varname, value):
             self._vars_cache[host] = combine_vars(self._vars_cache[host], {varname: value})
         else:
             self._vars_cache[host][varname] = value
+
+
+class VarsWithSources(dict):
```

Comment:
Although you have to implement a little more, use MutableMapping as the base class instead of dict.  (There's supposedly corner cases with dict)

Reply:
I know there are corner cases when subclassing `dict`, but I don't think any of them apply here. This passes all CI tests.